driver-core/src/main/com/mongodb/internal/operation/AsyncChangeStreamBatchCursor.java

```diff
@@ -227,23 +226,29 @@ public void onResult(final List<RawBsonDocument> rawDocuments, final Throwable t
         void apply(AsyncAggregateResponseBatchCursor<RawBsonDocument> cursor, SingleResultCallback<List<RawBsonDocument>> callback);
     }
 
-    private void resumeableOperation(final AsyncBlock asyncBlock, final SingleResultCallback<List<RawBsonDocument>> callback,
+    private void resumeableOperation(final AsyncBlock asyncBlock, final SingleResultCallback<List<RawBsonDocument>> convertResultsCallback,
                                      final boolean tryNext) {
         if (isClosed()) {
-            callback.onResult(null, new MongoException(format("%s called after the cursor was closed.",
+            convertResultsCallback.onResult(null, new MongoException(format("%s called after the cursor was closed.",
                     tryNext ? "tryNext()" : "next()")));
             return;
         }
-        asyncBlock.apply(getWrapped(), new SingleResultCallback<List<RawBsonDocument>>() {
+        AsyncAggregateResponseBatchCursor<RawBsonDocument> wrappedCursor = getWrapped();
+        asyncBlock.apply(wrappedCursor, new SingleResultCallback<List<RawBsonDocument>>() {
             @Override
             public void onResult(final List<RawBsonDocument> result, final Throwable t) {
+                boolean retry = false;
                 if (t == null) {
-                    callback.onResult(result, null);
+                    convertResultsCallback.onResult(result, null);
                 } else if (isRetryableError(t, maxWireVersion)) {
-                    nullifyAndCloseWrapped();
-                    retryOperation(asyncBlock, callback, tryNext);
+                    retry = true;
                 } else {
-                    callback.onResult(null, t);
+                    convertResultsCallback.onResult(null, t);
+                }
+                cachePostBatchResumeToken(wrappedCursor);
```

Comment:
Cashing after converting results and calculating `resumeToken` from them just as it's done in the synchronous counterpart (`ChangeStreamBatchCursor`). Note also that we must cache `resumeToken` before we retry.

Reply:
> As a result, postBatchResumeToken always overrides the resumeToken calculated from the received change stream documents if postBatchResumeToken is sent by the server (not sure if this is important, but this is how ChangeStreamBatchCursor is written).

:+1: `postBatchResumeToken` only started being sent by the server in MongoDB 4.0.7. Changes streams for 
rarely updated collections on a busy databases were failing the retryability, as the last seen resumeToken fell out the oplog. So sending a new token each getMore was the workaround.